granary/bluesky.py

```diff
@@ -461,12 +460,10 @@ def to_as1(obj, type=None):
     raise ValueError('Bluesky object missing $type field')
 
   # TODO: once we're on Python 3.10, switch this to a match statement!
-  if type in ('app.bsky.actor.defs#profileView',
-              'app.bsky.actor.defs#profileViewBasic'):
-    images = [{'url': obj.get('avatar')}]
-    banner = obj.get('banner')
```

Comment:
I think I removed this because it's not used anywhere and it also doesn't parse properly- `image` is expected to be an object with a `url` member, not an array. I used `mastodon.pyL490` as reference

Reply:
Double checking if my understanding is actually correct here as I think I've been looking at the AS2 spec by mistake. Is image in AS on a 'person' object supposed to be an array, an object, or either? I notice a lot of the Bluesky tests assume an array and that seems to match the as.json in the test data but it doesn't match what the Mastodon silo does so I'm a bit confused!

Reply:
Hah yeah AS1 vs AS2 can be confusing. The two useful spec links for AS1 are https://activitystrea.ms/specs/json/schema/activity-schema.html and https://activitystrea.ms/specs/json/1.0/ . granary also extends/overloads AS1 a bit here and there when we've outgrown it, and we don't have an explicity catalog of those extensions.

In this case, technically AS1 says `image` is an object with a `url` field (https://activitystrea.ms/specs/json/1.0/#object , https://activitystrea.ms/specs/json/1.0/#media-link ), and/or a bare string URL (https://activitystrea.ms/specs/json/schema/activity-schema.html#rfc.section.4.6 ). We sometimes need an object to have multiple images, though, so we support both lists and singular objects. Examples:
* https://github.com/snarfed/granary/blob/main/granary/tests/testdata/note_with_composite_photo.as.json
* https://github.com/snarfed/granary/blob/main/granary/tests/testdata/note_with_multiple_photos.as.json

(AS2 does this too, all of the properties not marked "Functional" can be either singular or plural. Whee! https://www.w3.org/TR/activitystreams-vocabulary/#h-properties )

We use a few helpers to deal with single vs plural, eg:
* `util.get_list` always returns a list
* `util.get_list` returns the first value
* `as1.get_object` always returns an object, `as1.get_objects` always a list. bare string values get wrapped in an object as the `id` value
* `as1.object_urls` returns a de-duped list of the `url` (singular) + `urls` (plural) values
etc...

Sorry this isn't simpler!

Reply:
(Also AS1 is only used internally, so we get to be a bit looser with it. We used to use it externally with Google+ and OStatus, and maybe others, but those are all long dead.)

Reply:
Cool cool. I think I'll need to make a corresponding change in bridgy itself:

```diff
diff --git a/pages.py b/pages.py
index 7d4098a2..c17f6779 100644
--- a/pages.py
+++ b/pages.py
@@ -197,10 +197,12 @@ def user(site, id):
         r.response['content'] = f'{r.actor.get("displayName") or ""} {phrase}.'
 
       # convert image URL to https if we're serving over SSL
-      image_url = r.actor.setdefault('image', {}).get('url')
+      # account for fact image might be a list
+      image = util.get_first(r.actor, 'image', {})
+      image_url = image.get('url')
       if image_url:
-        r.actor['image']['url'] = util.update_scheme(image_url, request)
-
+        image['url'] = util.update_scheme(image_url, request)
+        r.actor['image'] = image
       # generate original post links
       r.links = process_webmention_links(r)
       r.original_links = [util.pretty_link(url, new_tab=True)
```

Reply:
Yes! You probably want `image_url = util.get_url(image)` to handle bare string values as well as objects, otherwise looks good.